lib/libzfs/libzfs_iter.c

```diff
@@ -57,6 +63,169 @@ zfs_iter_clones(zfs_handle_t *zhp, zfs_iter_f func, void *data)
 	return (0);
 }
 
+static char *zfs_types[] = { "filesystem", "snapshot", "volume", "pool",
+	"bookmark" };
+
+#define	ZFS_TYPE_COUNT	(sizeof (zfs_types) / sizeof (&zfs_types[0]))
+
+static nvlist_t *
+zfs_type_to_nvl(zfs_type_t type)
+{
+	nvlist_t *nvl = fnvlist_alloc();
+	int i;
+
+	for (i = 0; i < ZFS_TYPE_COUNT; i++)
+		if (type & (1 << i))
+			fnvlist_add_boolean(nvl, zfs_types[i]);
+
+	return (nvl);
+}
+
+/*
+ * Iterate over all children filesystems (stable)
+ */
+int
+zfs_iter_generic(libzfs_handle_t *hdl, const char *name, zfs_type_t type,
+    int64_t depth, zfs_iter_f func, void *data)
+{
+	zfs_cmd_t zc = {"\0"};
+	zfs_handle_t *nzhp;
+	nvlist_t *tnvl;
+	nvlist_t *opts;
+	int fildes[2];
+	zfs_pipe_record_t zpr;
+	int ret;
+
+	if (zcmd_alloc_dst_nvlist(hdl, &zc, 0) != 0)
+		return (-1);
+
+	ret = pipe(fildes);
+	if (ret == -1) {
+		switch (errno) {
+		default:
+			ret = zfs_standard_error(hdl, errno,
+			    dgettext(TEXT_DOMAIN,
+			    "cannot iterate filesystems"));
+			break;
+		}
+	}
+
+	opts = fnvlist_alloc();
+	switch (depth) {
+	case -1:
+		fnvlist_add_boolean(opts, "recurse");
+		break;
+	default:
+		if (depth < 0) {
+			fnvlist_free(opts);
+			return (-1);
+		}
+		fnvlist_add_uint64(opts, "recurse", depth);
+	}
+
+	tnvl = zfs_type_to_nvl(type);
+	fnvlist_add_nvlist(opts, "type", tnvl);
+
+	if ((ret = lzc_list(name, fildes[1], opts)) != 0) {
+		fnvlist_free(opts);
+		close(fildes[0]);
+		close(fildes[1]);
+		return (ret);
+	}
+	fnvlist_free(tnvl);
+	fnvlist_free(opts);
+
+	while ((ret = read(fildes[0], &zpr,
+	    sizeof (zfs_pipe_record_t))) == sizeof (uint64_t) &&
+	    zpr.zpr_data_size != 0 && zpr.zpr_err == 0) {
+		nvlist_t *nvl, *nvl_prop, *nvl_dds;
+		char *name;
+#ifdef  _LITTLE_ENDIAN
+		uint32_t size = (zpr.zpr_endian) ? zpr.zpr_data_size :
+		    BSWAP_32(zpr.zpr_data_size);
+#else
+		uint32_t size = (zpr.zpr_endian) ? BSWAP_32(zpr.zpr_data_size) :
+		    zpr.zpr_data_size;
+#endif
+		char *buf = umem_alloc(size, UMEM_NOFAIL);
```

Comment:
Gotcha. I was wondering about the use of umem_alloc with no umem_nofail_callback. zfs_alloc seems like it would be suitable here.